src/template.c

```diff
@@ -5815,13 +5821,17 @@ void TemplateInstance::semantic(Scope *sc, Expressions *fargs)
     Dsymbols *target_symbol_list;
     size_t target_symbol_list_idx = 0;
     //if (sc->scopesym) printf("3: sc is %s %s\n", sc->scopesym->kind(), sc->scopesym->toChars());
-    if (!tinst && sc->scopesym && sc->scopesym->members)
+    if (0 && !tinst && sc->scopesym && sc->scopesym->members)
     {
         /* A module can have explicit template instance and its alias
          * in module scope (e,g, `alias Base64Impl!('+', '/') Base64;`).
-         * When the module is just imported, compiler can assume that
+         * When the module is just imported, normally compiler can assume that
          * its instantiated code would be contained in the separately compiled
-         * obj/lib file (e.g. phobos.lib). So we can omit their semantic3 running.
+         * obj/lib file (e.g. phobos.lib).
+         * Bugzilla 2644: However, if the template is instantiated in both
+         * modules of root and non-root, compiler should generate its objcode.
```

Comment:
The trouble is that people tend to write code where A imports B and B imports A, and then the templates are instantiated over and over. Currently, needsCodegen() instantiates the template in both A and B. I believe this is the source of the slowdown.

I think Martin has a good idea that for such cases, the module with a name that is "less" than the other one gets to instantiate it.